independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ContextDataMap.java

```diff
@@ -0,0 +1,163 @@
+        }
+        return delegate;
+    }
+
+    // ** Implement methods from the Map interface **
+
+    @Override
+    public void clear() {
+        throw new UnsupportedOperationException("Not allowed to clear the context data map");
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        if (ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS.equals(key)) {
+            return true;
+        }
+        return getDelegateForRead().containsKey(key);
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        if (interceptorBindings.equals(value)) {
+            return true;
+        }
+        return getDelegateForRead().containsValue(value);
+    }
+
+    @Override
+    public Set<Map.Entry<String, Object>> entrySet() {
+        final AbstractMap.SimpleImmutableEntry<String, Object> firstEntry = new AbstractMap.SimpleImmutableEntry<>(
+                ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS, interceptorBindings);
+        if (delegate == null) {
+            return Collections.singleton(firstEntry);
+        } else {
+            HashSet entries = new HashSet(delegate.size() + 1);
+            entries.addAll(delegate.entrySet());
+            entries.add(firstEntry);
+            return entries;
+        }
+    }
+
+    @Override
+    public Object get(Object key) {
+        if (ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS.equals(key)) {
+            return interceptorBindings;
+        }
+        return getDelegateForRead().get(key);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return false;
+    }
+
+    @Override
+    public Set<String> keySet() {
+        if (delegate == null) {
+            return Collections.singleton(ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS);
+        } else {
+            HashSet set = new HashSet(delegate.size() + 1);
+            set.addAll(delegate.keySet());
+            set.add(ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS);
+            return set;
+        }
+    }
+
+    @Override
+    public void putAll(Map m) {
+        if (m.containsKey(ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS)) {
+            throw new IllegalArgumentException(
+                    "Not allowed to put key '" + ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS + "' in the context data map");
+        }
+        getDelegateForWrite(m.size()).putAll(m);
+    }
+
+    @Override
+    public Object remove(Object key) {
+        if (ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS.equals(key)) {
+            throw new IllegalArgumentException("Not allowed to remove key '" + ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS
+                    + "' from the context data map");
+        } else if (delegate == null) {
+            return null;
+        } else {
+            return delegate.remove(key);
+        }
+    }
+
+    @Override
+    public int size() {
+        return getDelegateForRead().size() + 1;
+    }
+
+    @Override
+    public Collection values() {
```

Comment:
Is there a reason to use so many raw types in this class? I get a lot of useless warnings in my IDE ;-)